identity-actor/src/didcomm/system_builder.rs

```diff
@@ -0,0 +1,167 @@
+use futures::FutureExt;
+use libp2p::core::Executor;
+use libp2p::dns::TokioDnsConfig;
+use libp2p::identity::Keypair;
+use libp2p::tcp::TokioTcpConfig;
+use libp2p::websocket::WsConfig;
+use libp2p::Transport;
+
+use crate::actor::Actor;
+use crate::actor::ActorRequest;
+use crate::actor::Error;
+use crate::actor::Result as ActorResult;
+use crate::actor::System;
+use crate::actor::SystemBuilder;
+use crate::actor::SystemState;
+use crate::didcomm::AbstractDidCommActor;
+use crate::didcomm::ActorIdentity;
+use crate::didcomm::DidCommActor;
+use crate::didcomm::DidCommActorMap;
+use crate::didcomm::DidCommActorWrapper;
+use crate::didcomm::DidCommRequest;
+use crate::didcomm::DidCommSystem;
+use crate::didcomm::DidCommSystemState;
+use crate::p2p::EventLoop;
+use crate::p2p::InboundRequest;
+use crate::p2p::NetCommander;
+
+/// A builder for [`DidCommSystem`]s that allows for customizing its configuration and attaching actors.
+pub struct DidCommSystemBuilder {
+  inner: SystemBuilder,
+  identity: Option<ActorIdentity>,
+  didcomm_actors: DidCommActorMap,
+}
+
+impl DidCommSystemBuilder {
+  /// Create a new builder in the default configuration.
+  pub fn new() -> DidCommSystemBuilder {
+    Self {
+      inner: SystemBuilder::new(),
+      identity: None,
+      didcomm_actors: HashMap::new(),
+    }
+  }
+
+  /// See [`SystemBuilder::keypair`].
+  #[must_use]
+  pub fn keypair(mut self, keypair: Keypair) -> Self {
+    self.inner.keypair = Some(keypair);
+    self
+  }
+
+  /// Sets the timeout for [`DidCommSystem::await_message`] and the underlying libp2p
+  /// [`RequestResponse`](libp2p::request_response::RequestResponse) protocol.
+  #[must_use]
+  pub fn timeout(mut self, timeout: Duration) -> Self {
+    self.inner.config.timeout = timeout;
+    self
+  }
+
+  /// Set the [`ActorIdentity`] that will be used for DIDComm related tasks, such as en- and decryption.
+  #[must_use]
+  pub fn identity(mut self, identity: ActorIdentity) -> Self {
+    self.identity = Some(identity);
+    self
+  }
+
+  /// Attaches a [`DidCommActor`] to this system.
+  ///
+  /// This means that when the system receives a request of type `REQ`, it will invoke this actor.
+  ///
+  /// Calling this method multiple times with requests that have the same `Endpoint`
+  /// will detach the previous actor.
+  pub fn attach_didcomm<REQ, ACT>(&mut self, actor: ACT)
+  where
+    ACT: DidCommActor<REQ> + Send + Sync,
+    REQ: DidCommRequest + Send + Sync,
+  {
+    self.didcomm_actors.insert(
+      REQ::endpoint(),
+      Box::new(DidCommActorWrapper::new(actor)) as Box<dyn AbstractDidCommActor>,
+    );
+  }
+
+  /// Attaches an [`Actor`] to this system.
+  ///
+  /// This means that when the system receives a request of type `REQ`, it will invoke this actor.
+  ///
+  /// Calling this method with a `REQ` type whose endpoint is already attached to an actor
+  /// will overwrite the previous attachment.
+  pub fn attach<REQ, ACT>(&mut self, actor: ACT)
```

Comment:
I think this method can be removed. As far as I can see the standard actors are discarded once the `DIDCommSystem` is built or am I wrong about that?

Reply:
No, this adds actors to the `inner` `SystemBuilder`, which is later used to build a `System` which is then part of `DidCommSystem`. The `test_didcomm_system_supports_actor_requests` tests for exactly that.